### ISSUE TITLE
When focus is on an input box, it shows up in front of the sprite library

### ISSUE THREAD

**Issue by morantsur** [LLK/scratch-gui#1129]
![image](https://user-images.githubusercontent.com/15109618/34362697-86d017a8-ea7e-11e7-8421-69c718bb7aa6.png)
||
V

![image](https://user-images.githubusercontent.com/15109618/34362691-7fb00988-ea7e-11e7-8fc5-a40875c7d310.png)

### Steps to Reproduce
Add a block with an input field, and click it to focus on it.
Then, open the sprite library.

**Comment by rachel-fenichel** [comment]
Similar to https://github.com/LLK/scratch-gui/issues/1081

Needs a call to `Blockly.hideChaff()`.

**Comment by rachel-fenichel** [comment]
Frequently reported in Scratch 3.0 feedback.

**Comment by paulkaplan** [comment]
I don't think we should have to call `hideChaff` when modals are opened, it seems like it is just an issue where our modals aren't z-indexed high enough to cover the other elements. This line in particular looks suspicious. 

https://github.com/LLK/scratch-blocks/blob/e88260920ded6ac291a1027c064c9a80aaae9e7e/core/css.js#L148

**Comment by rachel-fenichel** [comment]
Calling `hideChaff` disposes of the widget cleanly, by calling its dispose function.  Otherwise the widget remains open, with all of its event handlers bound and so on, until the next time the user does something that causes scratch-blocks to call `hideChaff`.

For instance, if you have a text input field open, calling `hideChaff` will correctly call the function returned by `Blockly.FieldTextInput.prototype.widgetDispose_`, which does things like unbinding the keyboard listeners and saving the most recent text edit.  These seem really important.

Separately, the high z-index of the widget div may be suspicious.  It was a fairly aggressive response to a few bug reports.

**Comment by paulkaplan** [comment]
@rachel-fenichel I thought about this a little more.

So I realized it isn't possible to simply set the z-index lower, below our modals, because we also have a scratch-blocks workspace _within_ a modal, looking at you custom procedures...

But stepping back, it is unusual that the (html) input is still visible at all in these cases, because you have always clicked on another button to open a modal / go to fullscreen mode, which should mean that the (html) input is blurred. I would have thought that the editor was hidden on blur, but it doesn't seem to be. 

![blur-doesnt-close-editor](https://user-images.githubusercontent.com/654102/40132129-e65234aa-5909-11e8-994a-2eb334c3c4c3.gif)

(Note how the input is blurred, but the field editor remains)

Maybe instead of making other parts of the app call `Blockly.hideChaff`, we could make the text input call `WidgetDiv.hide()` the same way hitting <kbd>Enter</kbd> does, but on the `blur` event. That would hide the editor when you click one of the buttons that brings up a modal, making the whole field behave more like a standard html input. 

But now I may have talked myself in a circle, because for certain fields that extend the text input field, like the angle picker, there is extra UI in addition to the input. Those would not want to hide the editor when the text input is blurred. Perhaps those could override the onBlur to do nothing, but then we have the same problem but now for those fields :( 

I'll look into what it would take to make a "click-outside" wrapper for ScratchBlocks inside the GUI that can always "blur" (i.e. hideChaff) when you click something that isn't within the blocks. hmmm
